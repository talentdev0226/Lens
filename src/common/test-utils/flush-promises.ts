/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { setImmediate } from "timers";

export const flushPromises = () => new Promise(setImmediate);
