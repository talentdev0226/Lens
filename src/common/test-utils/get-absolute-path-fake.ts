/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { GetAbsolutePath } from "../path/get-absolute-path.injectable";

export const getAbsolutePathFake: GetAbsolutePath = (...args) => {
  const maybeAbsolutePath = args.join("/");

  if (isAbsolutePath(maybeAbsolutePath)) {
    return maybeAbsolutePath;
  }

  return `/some-absolute-root-directory/${maybeAbsolutePath}`;
};

const isAbsolutePath = (path: string) => path.startsWith("/");
