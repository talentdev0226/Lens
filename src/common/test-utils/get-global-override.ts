/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Injectable } from "@ogre-tools/injectable";

export interface GlobalOverride {
  injectable: Injectable<any, any, any>;
  overridingInstantiate: any;
}

export const getGlobalOverride = <T extends Injectable<any, any, any>>(
  injectable: T,
  overridingInstantiate: T["instantiate"],
) => ({
    injectable,
    overridingInstantiate,
  });
