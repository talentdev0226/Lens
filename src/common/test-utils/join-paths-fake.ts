/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { JoinPaths } from "../path/join-paths.injectable";

export const joinPathsFake: JoinPaths = (...args) => args.join("/");
