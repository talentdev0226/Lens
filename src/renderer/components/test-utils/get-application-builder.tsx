/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed, IObservableArray, observable, runInAction } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import { fireEvent, RenderResult } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import directoryForLensLocalStorageInjectable from "../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import { Sidebar } from "../layout/sidebar";
import { getDisForUnitTesting } from "../../../test-utils/get-dis-for-unit-testing";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import applicationMenuItemsInjectable, { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;
  addExtensions: (...extensions: LensRendererExtension[]) => Promise<ApplicationBuilder>;
  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeSetups: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigation: {
      click: (id: string) => void;
    };
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  renderSidebar: () => React.ReactNode;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const { rendererDi, mainDi, runSetups } = getDisForUnitTesting({
    doGeneralOverrides: true,
  });

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const extensionsState = observable.array<LensRendererExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments: Record<string, Environment> = {
    application: {
      renderSidebar: () => null,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    },

    clusterFrame: {
      renderSidebar: () => <Sidebar />,
      onAllowKubeResource: () => {},
    },
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => extensionsState),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed((): LensMainExtension[] => []),
  );

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click(undefined, undefined, undefined);
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join(
                '", "',
              )}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      rendererDi.override(
        directoryForLensLocalStorageInjectable,
        () => "/irrelevant",
      );

      return builder;
    },

    addExtensions: async (...extensions) => {
      const extensionRegistrators = rendererDi.injectMany(
        extensionRegistratorInjectionToken,
      );

      const addAndEnableExtensions = async () => {
        const registratorPromises = extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        );

        await Promise.all(registratorPromises);

        runInAction(() => {
          extensions.forEach((extension) => {
            extensionsState.push(extension);
          });
        });
      };

      if (rendered) {
        await addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }

      return builder;
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeSetups(callback: (dis: DiContainers) => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      for (const callback of beforeSetupsCallbacks) {
        await callback(dis);
      }

      await runSetups();

      const render = renderFor(rendererDi);

      const history = rendererDi.inject(observableHistoryInjectable);

      const currentRouteComponent = rendererDi.inject(
        currentRouteComponentInjectable,
      );

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      rendered = render(
        <Router history={history}>
          {environment.renderSidebar()}

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );

      return rendered;
    },
  };

  return builder;
};

const toFlatChildren =
  (parentId: string) =>
    ({
      submenu = [],
      ...menuItem
    }: MenuItemOpts): (MenuItemOpts & { path: string })[] =>
      [
        {
          ...menuItem,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        },
        ...submenu.flatMap(toFlatChildren(menuItem.id)),
      ];
