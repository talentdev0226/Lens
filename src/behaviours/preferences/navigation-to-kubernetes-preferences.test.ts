/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { RenderResult } from "@testing-library/react";
import { ApplicationBuilder, getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import userStoreInjectable from "../../common/user-store/user-store.injectable";
import type { UserStore } from "../../common/user-store";
import themeStoreInjectable from "../../renderer/theme-store.injectable";
import type { ThemeStore } from "../../renderer/theme.store";
import { observable } from "mobx";

describe("preferences - navigation to kubernetes preferences", () => {
  let applicationBuilder: ApplicationBuilder;

  beforeEach(() => {
    applicationBuilder = getApplicationBuilder();

    applicationBuilder.beforeSetups(({ rendererDi }) => {
      const userStoreStub = {
        extensionRegistryUrl: { customUrl: "some-custom-url" },
        syncKubeconfigEntries: observable.map(),
      } as unknown as UserStore;

      rendererDi.override(userStoreInjectable, () => userStoreStub);

      const themeStoreStub = ({ themeOptions: [] }) as unknown as ThemeStore;

      rendererDi.override(themeStoreInjectable, () => themeStoreStub);
    });
  });

  describe("given in preferences, when rendered", () => {
    let rendered: RenderResult;

    beforeEach(async () => {
      applicationBuilder.beforeRender(() => {
        applicationBuilder.preferences.navigate();
      });

      rendered = await applicationBuilder.render();
    });

    it("renders", () => {
      expect(rendered.container).toMatchSnapshot();
    });

    it("does not show kubernetes preferences yet", () => {
      const page = rendered.queryByTestId("kubernetes-preferences-page");

      expect(page).toBeNull();
    });

    describe("when navigating to kubernetes preferences using navigation", () => {
      beforeEach(() => {
        applicationBuilder.preferences.navigation.click("kubernetes");
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("shows kubernetes preferences", () => {
        const page = rendered.getByTestId("kubernetes-preferences-page");

        expect(page).not.toBeNull();
      });
    });
  });
});
