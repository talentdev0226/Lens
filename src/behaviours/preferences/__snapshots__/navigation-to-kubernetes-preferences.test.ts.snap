// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preferences - navigation to kubernetes preferences given in preferences, when rendered renders 1`] = `
<div>
  <div
    class="SettingLayout showNavigation Preferences"
    data-testid="application-preferences-page"
  >
    <nav
      class="sidebarRegion"
    >
      <div
        class="sidebar"
      >
        <div
          class="Tabs flex column"
        >
          <div
            class="header"
          >
            Preferences
          </div>
          <div
            class="Tab flex gaps align-center active"
            data-testid="tab-link-for-application"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              App
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-proxy"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Proxy
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-kubernetes"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Kubernetes
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-editor"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Editor
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-terminal"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Terminal
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="contentRegion"
      id="ScrollSpyRoot"
    >
      <div
        class="content"
      >
        <section
          id="application"
        >
          <h2
            data-testid="application-header"
          >
            Application
          </h2>
          <section
            id="appearance"
          >
            <div
              class="SubTitle"
            >
              Theme
               
            </div>
            <div
              class="Select theme-lens css-2b097c-container"
            >
              <div
                class="Select__control css-yk16xz-control"
              >
                <div
                  class="Select__value-container css-g1d714-ValueContainer"
                >
                  <div
                    class="Select__placeholder css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class="Select__input"
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="theme-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <hr />
          <section
            id="extensionRegistryUrl"
          >
            <div
              class="SubTitle"
            >
              Extension Install Registry
               
            </div>
            <div
              class="Select theme-lens css-2b097c-container"
            >
              <div
                class="Select__control css-yk16xz-control"
              >
                <div
                  class="Select__value-container css-g1d714-ValueContainer"
                >
                  <div
                    class="Select__placeholder css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class="Select__input"
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="extension-install-registry-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <p
              class="mt-4 mb-5 leading-relaxed"
            >
              This setting is to change the registry URL for installing extensions by name.
               
              If you are unable to access the default registry (
              https://registry.npmjs.org
              )
               
              you can change it in your 
              <b>
                .npmrc
              </b>
              Â file or in the input below.
            </p>
            <div
              class="Input theme round black disabled invalid"
            >
              <label
                class="input-area flex gaps align-center"
                id=""
              >
                <input
                  class="input box grow"
                  disabled=""
                  placeholder="Custom Extension Registry URL..."
                  spellcheck="false"
                  value="some-custom-url"
                />
              </label>
              <div
                class="input-info flex gaps"
              />
            </div>
          </section>
          <hr />
          <section
            id="other"
          >
            <div
              class="SubTitle"
            >
              Start-up
               
            </div>
            <label
              class=""
              data-testid="switch"
            >
              Automatically start Lens on login
              <input
                role="switch"
                type="checkbox"
              />
            </label>
          </section>
          <hr />
          <section
            id="update-channel"
          >
            <div
              class="SubTitle"
            >
              Update Channel
               
            </div>
            <div
              class="Select theme-lens css-2b097c-container"
            >
              <div
                class="Select__control css-yk16xz-control"
              >
                <div
                  class="Select__value-container css-g1d714-ValueContainer"
                >
                  <div
                    class="Select__placeholder css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class="Select__input"
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="update-channel-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <hr />
          <section
            id="locale"
          >
            <div
              class="SubTitle"
            >
              Locale Timezone
               
            </div>
            <div
              class="Select theme-lens css-2b097c-container"
            >
              <div
                class="Select__control css-yk16xz-control"
              >
                <div
                  class="Select__value-container css-g1d714-ValueContainer"
                >
                  <div
                    class="Select__placeholder css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class="Select__input"
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="timezone-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </section>
        </section>
      </div>
      <div
        class="toolsRegion"
      >
        <div
          class="fixed top-[60px]"
        >
          <div
            data-testid="close-preferences"
          >
            <div
              aria-label="Close"
              role="button"
            >
              <i
                class="Icon material focusable"
              >
                <span
                  class="icon"
                  data-icon-name="close"
                >
                  close
                </span>
              </i>
            </div>
            <div
              aria-hidden="true"
            >
              ESC
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`preferences - navigation to kubernetes preferences given in preferences, when rendered when navigating to kubernetes preferences using navigation renders 1`] = `
<div>
  <div
    class="SettingLayout showNavigation Preferences"
    data-testid="kubernetes-preferences-page"
  >
    <nav
      class="sidebarRegion"
    >
      <div
        class="sidebar"
      >
        <div
          class="Tabs flex column"
        >
          <div
            class="header"
          >
            Preferences
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-application"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              App
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-proxy"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Proxy
            </div>
          </div>
          <div
            class="Tab flex gaps align-center active"
            data-testid="tab-link-for-kubernetes"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Kubernetes
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-editor"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Editor
            </div>
          </div>
          <div
            class="Tab flex gaps align-center"
            data-testid="tab-link-for-terminal"
            role="tab"
            tabindex="0"
          >
            <div
              class="label"
            >
              Terminal
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="contentRegion"
      id="ScrollSpyRoot"
    >
      <div
        class="content"
      >
        <section
          id="kubernetes"
        >
          <section
            id="kubectl"
          >
            <h2
              data-testid="kubernetes-header"
            >
              Kubernetes
            </h2>
            <section>
              <div
                class="SubTitle"
              >
                Kubectl binary download
                 
              </div>
              <label
                class=""
                data-testid="switch"
              >
                Download kubectl binaries matching the Kubernetes cluster version
                <input
                  role="switch"
                  type="checkbox"
                />
              </label>
            </section>
            <section>
              <div
                class="SubTitle"
              >
                Download mirror
                 
              </div>
              <div
                class="Select theme-lens css-2b097c-container"
              >
                <div
                  class="Select__control css-yk16xz-control"
                >
                  <div
                    class="Select__value-container css-g1d714-ValueContainer"
                  >
                    <div
                      class="Select__placeholder css-1wa3eu0-placeholder"
                    >
                      Download mirror for kubectl
                    </div>
                    <div
                      class="css-b8ldur-Input"
                    >
                      <div
                        class="Select__input"
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="download-mirror-input"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </section>
            <section>
              <div
                class="SubTitle"
              >
                Directory for binaries
                 
              </div>
              <div
                class="Input theme round black disabled"
              >
                <label
                  class="input-area flex gaps align-center"
                  id=""
                >
                  <input
                    class="input box grow"
                    disabled=""
                    placeholder="/some-absolute-root-directory/some-electron-app-path-for-user-data/binaries"
                    spellcheck="false"
                    value=""
                  />
                </label>
                <div
                  class="input-info flex gaps"
                />
              </div>
              <div
                class="hint"
              >
                The directory to download binaries into.
              </div>
            </section>
            <section>
              <div
                class="SubTitle"
              >
                Path to kubectl binary
                 
              </div>
              <div
                class="Input theme round black"
              >
                <label
                  class="input-area flex gaps align-center"
                  id=""
                >
                  <input
                    class="input box grow"
                    placeholder="/some-absolute-root-directory/some-electron-app-path-for-user-data/binaries/kubectl"
                    spellcheck="false"
                    value=""
                  />
                </label>
                <div
                  class="input-info flex gaps"
                />
              </div>
            </section>
          </section>
          <hr />
          <section
            id="kube-sync"
          >
            <h2
              data-testid="kubernetes-sync-header"
            >
              Kubeconfig Syncs
            </h2>
            <div
              class="self-start mb-5"
            >
              <button
                class="Button PathPicker primary"
                type="button"
              >
                Sync Files and Folders
              </button>
            </div>
            <div
              class="SubTitle pt-5"
            >
              Synced Items
               
            </div>
            <div
              class="mt-3"
            >
              <div
                class="flex-grow text-center"
              >
                No files and folders have been synced yet
              </div>
            </div>
          </section>
          <hr />
          <section
            id="helm"
          >
            <h2>
              Helm Charts
            </h2>
            <div>
              <div
                class="flex gaps"
              >
                <div
                  class="Select theme-lens box grow css-2b097c-container"
                >
                  <div
                    class="Select__control css-yk16xz-control"
                  >
                    <div
                      class="Select__value-container css-g1d714-ValueContainer"
                    >
                      <div
                        class="Select__placeholder css-1wa3eu0-placeholder"
                      >
                        Repositories
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class="Select__input"
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="HelmRepoSelect"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-6q0nyr-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
                <button
                  class="Button primary"
                  type="button"
                >
                  Add Custom Helm Repo
                </button>
              </div>
              <div>
                <div
                  class=""
                >
                  <div
                    class="flex-grow text-center"
                  >
                    The repositories have not been added yet
                  </div>
                </div>
              </div>
            </div>
          </section>
        </section>
      </div>
      <div
        class="toolsRegion"
      >
        <div
          class="fixed top-[60px]"
        >
          <div
            data-testid="close-preferences"
          >
            <div
              aria-label="Close"
              role="button"
            >
              <i
                class="Icon material focusable"
              >
                <span
                  class="icon"
                  data-icon-name="close"
                >
                  close
                </span>
              </i>
            </div>
            <div
              aria-hidden="true"
            >
              ESC
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
